autoprecompiles/src/memory_optimizer.rs

```diff
@@ -11,19 +12,19 @@ use powdr_constraint_solver::runtime_constant::{RuntimeConstant, VarTransformabl
 use powdr_constraint_solver::utils::possible_concrete_values;
 use powdr_number::FieldElement;
 
-/// The memory address space for register memory operations.
-const REGISTER_ADDRESS_SPACE: u32 = 1;
+use crate::adapter::{Adapter, GroupedExpressionPowdr};
+use crate::expression::AlgebraicReference;
 
 /// Optimizes bus sends that correspond to general-purpose memory read and write operations.
 /// It works best if all read-write-operation addresses are fixed offsets relative to some
 /// symbolic base address. If stack and heap access operations are mixed, this is usually violated.
-pub fn optimize_memory<T: FieldElement, V: Hash + Eq + Clone + Ord + Display>(
-    mut system: ConstraintSystem<T, V>,
+pub fn optimize_memory<A: Adapter>(
```

Comment:
Could've added a `M: MemoryBusInteraction<_, _>` parameter here, but figured it's better to pass the adapter.

Reply:
If we decouple it from Adapter, we can still change the variable type easily. What this function here does is essentially returning a list of bus interaction indices to be removed and a set of substitutions / constraints. If we want to transform the constraint system before we apply memory optimization, we can probably apply those effects to the un-transformed system, but not if we are tied to the exact adapter.

Reply:
Yeah, don't have a strong opinion on this! I made it generic again in the memory optimizer:
e25953eaae6632ffd150017318bb6f03cf3319ac

Looks fine to me!

I left the changes to `autoprecompiles/src/optimizer.rs` though, because those are actually convenient in #3070 as well.

Reply:
Nice, thanks! Yeah, makes sense to keep the optimizer specific, it ties all the others together and if we make such a change as I said above, it would be in the optimizer.